### ISSUE TITLE
Investigate banning nullable enums

### ISSUE THREAD

**Issue by annevk** [heycam/webidl#213]
I cannot think of a single scenario where that would be preferable over an additional enum value. Of course, it might be too late to change this.

**Comment by foolip** [comment]
Should optional dictionary members of enum type also be required to have a default value? Otherwise undefined can mean something other than one of the enum values.

**Comment by annevk** [comment]
I think that would break `Request` objects.

**Comment by bzbarsky** [comment]
Optional enum values (in dictionaries, and as arguments) may make sense as a "use the default" kind of thing, especially when the default is computed in a complicated way from other arguments or dictionary members, or depends on prior state or whatnot.

I looked at where nullable enums are used in Gecko's IDL.  Apart from codegen tests and some nonstandard APIs, they are not used.

In the nonstandard APIs, there are some cases in which null is used to indicate "object is not in a state that can provide this data right now".  And boy are there a lot of uses like that.  :(  Those APIs were all written by JS hackers, for what it's worth, so it's a data point in terms of what "authors" expect.

**Comment by foolip** [comment]
I found a nullable enum in Blink and Gecko:
https://chromium.googlesource.com/chromium/src/+/master/third_party/WebKit/Source/modules/peerconnection/RTCSessionDescription.idl
https://github.com/mozilla/gecko-dev/blob/master/dom/webidl/RTCSessionDescription.webidl

Found via https://crbug.com/662898 because it doesn't match the spec.

**Comment by bzbarsky** [comment]
Yes, I found that too and filed https://bugzilla.mozilla.org/show_bug.cgi?id=1313966 on Gecko at the time.  I didn't include it in https://github.com/heycam/webidl/issues/213#issuecomment-257209695 as a result.

**Comment by foolip** [comment]
Do you recall if there were any others that are web exposed and could be risky to change?

**Comment by bzbarsky** [comment]
There aren't any.

**Comment by foolip** [comment]
Great, thanks for checking!

**Comment by foolip** [comment]
Here's a new one: https://wicg.github.io/media-capabilities/#screen-extension

The `colorGamut` attribute is a nullable enum. In this case, I suppose that the empty string could be used instead, but would that be an improvement? Note that `luminance` is also nullable but of interface type.

**Comment by foolip** [comment]
@mounirlamouri ^

**Comment by annevk** [comment]
Yeah, I don't see why using two types is better than a single type for `colorGamut`.

**Comment by tobie** [comment]
> I suppose that the empty string could be used instead, but would that be an improvement?

From a JS dev perspective, empty strings as falsy values in enums are icky (see `canPlay`).

**Comment by annevk** [comment]
@tobie that's also an argument against null though.

**Comment by tobie** [comment]
> that's also an argument against null though.

Well, not really. `null` in that case pretty much means: "object is not in a state that can provide this data right now" as @bzbarsky mentioned above, or "this has not be set yet." The empty string on the other hand is just super weird (and hacky in a way I wouldn't want to see Web APIs hacky). So if I had no other consideration in mind than "how does that feel to a Web dev," I'd say `null` is OK, but `""` should be banned.

**Comment by annevk** [comment]
Null is just as falsy though, so if you want to avoid that you should do something else.

**Comment by tobie** [comment]
> Null is just as falsy though, so if you want to avoid that you should do something else.

Well, all of this is pretty subjective anyway. The point I was trying to make is that replacing `null` by `""` isn't an improvement from a JS dev perspective. I'd even argue it's worse. So (still from a JS dev's perspective—as I said earlier, there might be other sound reasons to choose differently), my suggestion would be to keep nullable enums (if those are needed) but to disallow empty strings in enums.

**Comment by foolip** [comment]
I noticed that `colorGamut` from https://github.com/heycam/webidl/issues/213#issuecomment-279397891 is no longer nullable.

**Comment by foolip** [comment]
Found some more nullable enums:
https://drafts.csswg.org/web-animations/#dictdef-basecomputedkeyframe
https://drafts.csswg.org/web-animations/#dictdef-basepropertyindexedkeyframe
https://drafts.csswg.org/web-animations/#dictdef-basekeyframe
https://w3c.github.io/web-nfc/#dom-nfcwatchoptions
https://w3c.github.io/webrtc-pc/#dom-rtcicecandidate
https://w3c.github.io/webrtc-pc/#dom-rtcrtptransceiver

**Comment by foolip** [comment]
@alexshalamov, can you comment on the Web NFC case? Since it's an optional dictionary member, I believe just using undefined to mean missing would be fine. I've filed https://github.com/w3c/web-nfc/issues/149.

**Comment by birtles** [comment]
> https://drafts.csswg.org/web-animations/#dictdef-basecomputedkeyframe
>     https://drafts.csswg.org/web-animations/#dictdef-basepropertyindexedkeyframe
>     https://drafts.csswg.org/web-animations/#dictdef-basekeyframe
> 
> Yeah, this one is a bit weird. It might have been nicer to have an explicit "default" value or something... @birtles do you recall why null was picked here? I should have asked this in https://bugzilla.mozilla.org/show_bug.cgi?id=1429671...

We could have a "default" value, I guess. The reasoning at the time is mostly covered in https://github.com/w3c/csswg-drafts/issues/2072#issuecomment-349195346, where consistency with `offset`, which uses `null` for "auto"-like values, seems to be the main reason. (The second reason listed there concerns whether we represent auto-values by lack of a value or an explicit sentinel value--but either `null` or "default" would work for that).

That is you can write:

```js
elem.animate({
  backgroundColor: ['red', 'green', 'blue', 'orange', 'purple'],
  offset: [0, null, null, 0.95, 1],
}, 1000);
```

(where `null` means, "Calculate the offset for me").

So, it made sense that `composite` should use the same notation (where `null`
means, "Use the composite mode specified on the effect"):

i.e.

```js
elem.animate({
  backgroundColor: ['red', 'green', 'blue', 'orange', 'purple'],
  offset: [0, null, null, 0.95, 1],
  composite: ['replace', null, null, null, 'replace'],
}, 1000);
```

With a default value that would become:

```js
elem.animate({
  backgroundColor: ['red', 'green', 'blue', 'orange', 'purple'],
  offset: [0, null, null, 0.95, 1],
  composite: ['replace', 'auto', 'auto', 'auto', 'replace'],
}, 1000);
```

Which could work.

We haven't shipped composite modes in any browser but we are planning to ship
`getKeyframes()` in Firefox 63 which exposes `BaseComputedKeyframe`, i.e. will
return objects with `composite` set to `null`. However we'd still be in time to
change this if we want to.

**Comment by marcoscaceres** [comment]
“auto” looks way more legible and obvious, fwiw. I’d encourage you to change it.

**Comment by birtles** [comment]
Ok, I've filed a [PR to change Web animations](https://github.com/w3c/csswg-drafts/pull/3001) and written a [patch for Firefox and wpt](https://hg.mozilla.org/try/rev/a1db859c68ea0c2b0e6f2b752563de6ed13115e2) should that PR be accepted.

**Comment by birtles** [comment]
I've updated Web Animations (and Firefox implementation) so that it no longer uses nullable enums.
